src/os/bluestore/BlueStore.cc

```diff
@@ -5475,71 +5534,149 @@ int BlueStore::_write_out_fm_meta(uint64_t target_size)
+    shared_alloc.set(alloc);
+  }
+
   return 0;
 }
 
-int BlueStore::_init_alloc()
+int BlueStore::_init_alloc(bool read_only)
 {
   int r = _create_alloc();
   if (r < 0) {
     return r;
   }
-  ceph_assert(shared_alloc.a != NULL);
+  ceph_assert(alloc != NULL);
 
 #ifdef HAVE_LIBZBD
   if (bdev->is_smr()) {
-    auto a = dynamic_cast<ZonedAllocator*>(shared_alloc.a);
+    auto a = dynamic_cast<ZonedAllocator*>(alloc);
     ceph_assert(a);
     auto f = dynamic_cast<ZonedFreelistManager*>(fm);
     ceph_assert(f);
-    a->init_alloc(f->get_zone_states(db),
-		  &zoned_cleaner_lock,
-		  &zoned_cleaner_cond);
+    vector<uint64_t> wp = bdev->get_zones();
+    vector<zone_state_t> zones = f->get_zone_states(db);
+    ceph_assert(wp.size() == zones.size());
+
+    // reconcile zone state
+    auto num_zones = bdev->get_size() / zone_size;
+    map<uint32_t, uint64_t> zone_adjustments;
+    for (unsigned i = first_sequential_zone; i < num_zones; ++i) {
+      uint64_t p = wp[i] == (i + 1) * zone_size ? zone_size : wp[i] % zone_size;
+      if (zones[i].write_pointer > p) {
+	derr << __func__ << " zone 0x" << std::hex << i
+	     << " bluestore write pointer 0x" << zones[i].write_pointer
+	     << " > device write pointer 0x" << p
+	     << std::dec << " -- VERY SUSPICIOUS!" << dendl;
+      } else if (zones[i].write_pointer < p) {
+	// this is "normal" in that it can happen after any crash (if we have a
+	// write in flight but did not manage to commit the transaction)
+	auto delta = p - zones[i].write_pointer;
+	dout(1) << __func__ << " zone 0x" << std::hex << i
+		 << " device write pointer 0x" << p
+		 << " > bluestore pointer 0x" << zones[i].write_pointer
+		 << ", advancing 0x" << delta << std::dec << dendl;
+	zone_adjustments[i] = delta;
+	zones[i].num_dead_bytes += delta;
+	zones[i].write_pointer = p;
+      }
+    }
+
+    // start with conventional zone "free" (bluefs may adjust this when it starts up)
+    auto reserved = _get_ondisk_reserved();
+    shared_alloc.a->init_add_free(reserved,
+				  bdev->get_conventional_region_size() - reserved);
+
+    // init sequential zone based on the device's write pointers
+    a->init_from_zone_pointers(zones);
+    dout(1) << __func__
+	    << " loaded zone pointers: "
+	    << std::hex
+	    << ", allocator type " << alloc->get_type()
+	    << ", capacity 0x" << alloc->get_capacity()
+	    << ", block size 0x" << alloc->get_block_size()
+	    << ", free 0x" << alloc->get_free()
+	    << ", fragmentation " << alloc->get_fragmentation()
+	    << std::dec << dendl;
+
+    if (!read_only && !zone_adjustments.empty()) {
+      dout(1) << __func__ << " adjusting freelist based on device write pointers" << dendl;
+      KeyValueDB::Transaction t = db->get_transaction();
+      for (auto& i : zone_adjustments) {
+	// note that the offset is imprecise, but only need to select the zone
+	f->allocate(i.first * zone_size, i.second, t);
+      }
+      db->submit_transaction_sync(t);
+    }
```

Comment:
We invoke `_init_alloc()` from `_open_db_and_around()`, in the part where db is always opened in read-only.
The value _read_only_ is misleading at this point - when *false* it just means that in future we will reopen db in r/w mode.

Reply:
the code as it stands is correct, though, right?  the first (read-only) pass won't adjust the freelist's write pointer, but the second (!read-only) pass will?

Reply:
To be sure I added test.
https://github.com/ceph/ceph/pull/43257
It fails on DB when write pointer needs to be fixed.

Reply:
Ah, I see it now!  Thanks for that test